# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric

In [2]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [3]:
x.shape

(506, 13)

In [4]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [5]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 

    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists
    selecting_rows = np.random.choice(len(input_data), size=303, replace=False) 
    replicating_rows = np.random.choice(len(selecting_rows),size= 203,replace=True)
    col = np.arange(3,len(input_data[1]),1)
    no_of_col = np.random.choice(col)
    range_col=  [*range(0,13,1)]
    selecting_columns = np.random.choice(range_col,size=no_of_col, replace=False)
    sample_data = input_data[selecting_rows[:, None],selecting_columns]
    target_sample_data = target_data[selecting_rows]
    ############ 
    # replicating sample data
    replicating_sample_data = sample_data[replicating_rows]
    target_replicated_sample_data = target_data[replicating_rows]
    ###########
    # concatnating data
    final_sample_data = np.vstack((sample_data,replicating_sample_data))
    final_target_data = np.vstack((target_sample_data.reshape(-1,1),target_replicated_sample_data.reshape(-1,1)))
    
    return list(final_sample_data),list(final_target_data),list(selecting_rows),list(selecting_columns)

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [6]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [7]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]
for i in range(0,30):
    a,b,c,d = generating_samples(x,y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [8]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [9]:
from sklearn.tree import DecisionTreeRegressor
input_data =[]
output_data=[]
list_of_models=[]
for i in range(30):
    input_data.append(x[np.asarray(list_selected_row[i])[:,None], np.asarray(list_selected_columns[i])[None,:]])
    output_data.append(y[np.asarray(list_selected_row[i])[:,None]])
    model = DecisionTreeRegressor(max_depth=None)
    model.fit(input_data[i],output_data[i])
    list_of_models.append(model)

In [10]:
list_of_models

[DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                       max_leaf_nodes=None, min_impurity_decrease=0.0,
                       min_impurity_split=None, min_samples_leaf=1,
                       min_samples_split=2, min_weight_fraction_leaf=0.0,
                       presort=False, random_state=None, splitter='best'),
 DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                       max_leaf_nodes=None, min_impurity_decrease=0.0,
                       min_impurity_split=None, min_samples_leaf=1,
                       min_samples_split=2, min_weight_fraction_leaf=0.0,
                       presort=False, random_state=None, splitter='best'),
 DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                       max_leaf_nodes=None, min_impurity_decrease=0.0,
                       min_impurity_split=None, min_samples_leaf=1,
                       min_samples_split=2, min_weight_fract

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [11]:
y_pred=[]
for i in range(506):
    array_for_y =[]
    for j in range(30):
        array_for_y.append(list_of_models[j].predict(x[i][list_selected_columns[j]].reshape(1,-1)))
    y_pred.append(np.median(array_for_y))


print("The MSE:")
diff=[]
for i in range(506):
    diff.append((y[i]-y_pred[i])**2)
print(np.sum(diff)/506)
    

The MSE:
0.03526679841897236


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [12]:
import warnings
warnings.filterwarnings("ignore")
y_pred1=[]
for i in range(506):
    array_for_y_new =[]
    for j in range(30):
        if i not in list_selected_row[j]:
            array_for_y_new.append(list_of_models[j].predict(x[i][list_selected_columns[j]].reshape(1,-1)))  
    y_pred1.append(np.median(array_for_y_new))


print("The OOB Score:")
diff1=[]
for i in range(506):
    diff1.append((y[i]-y_pred1[i])**2)
print(np.sum(diff1)/506)

The OOB Score:
16.288305335968378


# <font color='blue'><b>Task 2</b></font>

In [13]:
import tqdm
mse=[];oob=[];
for i in tqdm.tqdm(range(35)):
    list_input_data =[]
    list_output_data =[]
    list_selected_row= []
    list_selected_columns=[]
    for i in range(0,30):
        a,b,c,d = generating_samples(x,y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)
    input_data =[]
    output_data=[]
    list_of_models=[]
    for i in range(30):
        input_data.append(x[np.asarray(list_selected_row[i])[:,None], np.asarray(list_selected_columns[i])[None,:]])
        output_data.append(y[np.asarray(list_selected_row[i])[:,None]])
        model = DecisionTreeRegressor(max_depth=None)
        model.fit(input_data[i],output_data[i])
        list_of_models.append(model)
    y_pred=[]
    for i in range(506):
        array_for_y =[]
        for j in range(30):
            array_for_y.append(list_of_models[j].predict(x[i][list_selected_columns[j]].reshape(1,-1)))
        y_pred.append(np.median(array_for_y))
    warnings.filterwarnings("ignore")
    y_pred1=[]
    for i in range(506):
        array_for_y_new =[]
        for j in range(30):
            if i not in list_selected_row[j]:
                array_for_y_new.append(list_of_models[j].predict(x[i][list_selected_columns[j]].reshape(1,-1)))  
        y_pred1.append(np.median(array_for_y_new))
    mse.append(mean_squared_error(y,y_pred))
    oob.append(mean_squared_error(y,y_pred1))
print("MSE:",mse)
print('#'*100)
print("OOB:",oob)

100%|██████████████████████████████████████████| 35/35 [01:42<00:00,  2.93s/it]


MSE: [0.22413921826965313, 0.026246603260869566, 0.07922430830039523, 0.011329051383399194, 0.15234574000878343, 0.028379446640316247, 0.4535140810276678, 0.11630327371693992, 0.1967008399209486, 0.29019762845849806, 0.09386487154150201, 0.15107240338164263, 0.36606952129995607, 0.11937005928853762, 0.035454545454545516, 0.1347776679841897, 0.1894577569169961, 0.07993083003952571, 0.21189723320158108, 0.03809288537549412, 0.0973381916996047, 0.026541501976284597, 0.033414031620553376, 0.016679841897233202, 0.1352371541501976, 0.06391921936758896, 0.062223320158102655, 0.3039538043478262, 0.0008053359683794527, 0.08194883618796664, 0.5897903285573122, 0.05143280632411064, 0.06268774703557313, 0.04146739130434788, 0.12012957393925955]
####################################################################################################
OOB: [13.883775563287035, 13.884888833992095, 14.988275828941587, 13.944403958058846, 17.343254861252984, 20.776191509861313, 16.308720355731225, 12.5648462

### Calculating Confidence Interval CI

In [14]:
import math
sample_mean_mse = np.mean(mse)
sample_std_mse = np.std(mse)
n = len(mse)
standard_error_mean_mse = sample_std_mse/(math.sqrt(n))
print("The Confidence Interval of MSE:")
print('[',sample_mean_mse-(2*standard_error_mean_mse),',',sample_mean_mse+(2*standard_error_mean_mse),']')


The Confidence Interval of MSE:
[ 0.0899347151986333 , 0.17783311623026857 ]


In [15]:
sample_mean_oob = np.mean(oob)
sample_std_oob = np.std(oob)
m = len(oob)
standard_error_mean_oob = sample_std_oob/(math.sqrt(m))
print("The Confidence Interval of OOB:")
print('[',sample_mean_oob-(2*standard_error_mean_oob),',',sample_mean_oob+(2*standard_error_mean_oob),']')

The Confidence Interval of OOB:
[ 14.32721837538061 , 15.775525827409282 ]


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [16]:
xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
warnings.filterwarnings("ignore")
yq=[]
array_for_y_q =[]
for j in range(30):
    if i not in list_selected_row[j]:
         array_for_y_q.append(list_of_models[j].predict(np.array(xq)[list_selected_columns[j]].reshape(1,-1)))  
yq.append(np.median(array_for_y_q))
print('The predicted value for the given xq: ',float(np.asarray(yq)))

The predicted value for the given xq:  18.9


<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

#### TASK 1: 
The Boston Dataset is used here which has prices of houses from various locations. It contains 506 rows and 13 features(crime-rate,age,etc) <br /> 
Here I apply Bootstrapping method and then use DecisionTreeRegressor as our model. <br /> 
By using the function "generating_samples" we obtain a sample dataset from the actual given data. <br /> 
The function returns a sample data of size 506 with number of columns ranging from 3 to 13. <br /> 
We create 30 samples ensuring each sample has different set of columns. <br /> 
<br /> 

By using DecisionTreeRegressor() models fit on input data column sampled on each of the 30 samples, <br /> 
I found MSE = 0.03526679841897236 <br /> 
and Out of Bag(oob) Score = 16.288305335968378 <br /> 
<br /> 

#### TASK 2:
Here, I generate 35 samples and obtain 35 MSE and OOB values in the form of a list. <br /> 

Next I calculate the Confidence Intervals of MSE & OOB Score taking above two lists as samples.<br /> 
By refering the Central_limit_theorem notebook, <br /> 
since we have no population std_dev statistic, we apply the case 2 of the conclusion given in above mentioned notebook. <br /> 
I have obtained: <br /> 
The Confidence Interval of MSE: <br /> 
[ 0.0899347151986333 , 0.17783311623026857 ] <br /> 
The Confidence Interval of OOB: <br /> 
[ 14.32721837538061 , 15.775525827409282 ] <br /> 
<br /> 

#### TASK 3: 

Using 30 models by using 30 samples in TASK-1, the predicted output for the given xq was obtained to be 18.9